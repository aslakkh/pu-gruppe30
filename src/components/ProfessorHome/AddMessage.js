/**
 * Created by jan on 21/03/2017.
 */
import React, { Component } from 'react'
import firebase from 'firebase';
import {ref } from '../../config/constants'
import {Form, FieldGroup, FormControl} from 'react-bootstrap'


/*
* Lets the professor add a message the student can see
 */

export default class AddMessage extends Component {
    constructor(props){
        super(props);
        this.state={
            courseID:this.props.courseID,
        }
    }
    handleSubmit = (e) =>{
        e.preventDefault();
        this.castMessageToFirebase()

    }
    /*
    * Send the message to firebase and uses timestap as ID
    */
     castMessageToFirebase(){
         const userUid = firebase.auth().currentUser.uid;
         console.log(this.state.courseID)
         console.log(this.message.value)
         console.log('courses/'+this.state.courseID+'/Messages/'+ Date.now())
         const timeRef = ref.child('courses/'+this.state.courseID+'/Messages/'+ Date.now());
         timeRef.set({
             Message: this.message.value
         })
         this.message.value=""


    }


    render(){
        return(


            <form onSubmit={this.handleSubmit}>
                <FormControl
                    id="formControlsText"
                    type="text"
                    inputRef={ (ref) => {this.message = ref;}}
                    label="Message"
                    placeholder="Message"
                />
                <button type="submit" className="btn btn-primary">Submit</button>

            </form>

        )
    }


}