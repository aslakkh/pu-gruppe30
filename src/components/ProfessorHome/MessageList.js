/**
 * Created by jan on 21/03/2017.
 */
import React, { Component } from 'react'
import firebase from 'firebase';
import {ref } from '../../config/constants'
import {Form, FieldGroup, FormControl, ListGroup, ListGroupItem, Button} from 'react-bootstrap'


/*
* Displays list of messages and delete selected messages.
*/

export default class MessageList extends Component {
    constructor(props){
        super(props)
        this.state={
            courseID: this.props.courseID,
            messages:undefined
        }
    }

    getCommentsFromFirebase(){
        let messageRef = firebase.database().ref();
        let that = this;
        var messages =[]
        var koder = []
        messageRef.child('courses/'+this.state.courseID+'/Messages').on('value', function(snapshot) {
            snapshot.forEach(function(data){
                messages.unshift([ data.key,data.val().Message])

            })
            that.setState({
                messages:messages
            })
            messages=[]
    })
        console.log(this.state.messages)
    }

    componentWillMount(){
        this.getCommentsFromFirebase()
    }


handleClick(key) {
    var userRef = ref.child('courses/' + this.state.courseID + '/Messages/');
    userRef.child(key).remove();
}
    render(){

        if(this.state.messages === undefined || this.state.messages === null){
            return(
                <h5>You have no active courses. Add a course below.</h5>
            );
        }
        else{
            return(
                <div>

                    <ListGroup >
                        {Object.keys(this.state.messages).map((key,i) => {
                            return <ListGroupItem key={key} className="CoursesList">
                                {this.state.messages[key][1]}

                                <Button bsStyle="danger" onClick={() => this.handleClick(this.state.messages[key][0])}>Delete</Button>
                            </ListGroupItem>
                        })}
                    </ListGroup>
                </div>
            );
        }


    }
}