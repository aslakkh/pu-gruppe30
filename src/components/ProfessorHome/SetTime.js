import React, { Component } from 'react';
import {Button, Modal, FormGroup, Label, Form, DropdownButton, MenuItem} from 'react-bootstrap';
import { ref } from '../../config/constants'
import {getDaysHoursMins} from '../../helpers/helperFunctions';
import { setRecommendedTime } from '../../helpers/auth'
import FeedbackMessage from '../FeedbackMessage'


export default class SetTime extends Component {

    constructor(props) {
        super(props);
        this.handleSaveBtn = this.handleSaveBtn.bind(this);
        this.open = this.open.bind(this);
        this.close = this.close.bind(this);
        this.state = {
            courseID: this.props.courseID,
            daySelected: '0',
            hourSelected: '0',
            minSelected: '0',
            view: "Weekly Goal",
        }
    }

    componentWillReceiveProps(nextProps) {
        
        //remove feedbackmessage (if any) and update course and time values if user has moved to a different course page
        if(nextProps.courseID !== this.state.courseID){
            this.setState({
                displayFeedbackMessage: false,
                courseID: nextProps.courseID,
                daySelected: '0',
                hourSelected: '0',
                minSelected: '0',
            })
        }
        
    }
    


    handleSaveBtn(e){
        /*
         On save-click dropdown values are converted into seconds and stored in firebase
         */
        e.preventDefault(); //prevents default browser behaviour on click
        let seconds = 0;

        //convert input to seconds
        seconds = (parseInt(this.state.minSelected) * 60) + (parseInt(this.state.hourSelected) * 60 * 60) + (parseInt(this.state.daySelected) * 12 * 60 * 60);


        if (seconds > 0) {
            let view = "weeklyGoal";
            setRecommendedTime(this.state.courseID, seconds);
            this.setState({
                displayFeedbackMessage: true,
                feedbackMessage: "Recommended time per week for " + this.state.courseID + " set to " + this.state.daySelected + " days, " + this.state.hourSelected + " hours, " + this.state.minSelected + " minutes. ",
                bsStyle: 'success',
            });
        }
        else{
            this.setState({
                displayFeedbackMessage: true,
                feedbackMessage: "Invalid value: recommended time must be greater than 0. ",
                bsStyle: 'warning',
            });
        }
    };

    getDropdowns(view) {
        /*
         Helper method for render method. Populates the dropdownmenus with correct information.
         */
        let items = [];
        if (view === 'Monthly Goal' || view === 'Weekly Goal') {
            items.push(<Label className="plain-label" key={1}>Days:</Label>);
            items.push(<DropdownButton key={2} title={this.state.daySelected}
                                       id={'dropdown-basic'}>{this.getMenuItems('days', view)}</DropdownButton>);
        }
        items.push(<Label className="plain-label" key={3}>Hours:</Label>);
        items.push(<DropdownButton key={4} title={this.state.hourSelected}
                                   id={'dropdown-basic'}>{this.getMenuItems('hours', view)}</DropdownButton>);
        items.push(<Label className="plain-label" key={5}>Minutes:</Label>);
        items.push(<DropdownButton key={6} title={this.state.minSelected}
                                   id={'dropdown-basic'}>{this.getMenuItems('minutes', view)}</DropdownButton>);
        return (
            <Form className="enter-form" inline>
                {items}
            </Form>
        )
    }

    getMenuItems(box, view) {
        let items = [];
        let max = 0;
        view === "Monthly Goal" ? max=32 : view === 'Weekly Goal' ? max = 8 : 0;
        if (box==='days') {
            for(let i=0; i<max; i++) {
                items.push(<MenuItem onClick={() => this.setDays(i)} key={i}>{i}</MenuItem>);
            }
        }
        if (box==='hours') {
            for(let i=0; i<13; i++) {
                items.push(<MenuItem onClick={() => this.setHours(i)} key={i}>{i}</MenuItem>);
            }
        }
        if (box==='minutes') {
            for(let i=0; i<60; i+=15) {
                items.push(<MenuItem onClick={() => this.setMinutes(i)} key={i}>{i}</MenuItem>);
            }
        }
        return items;
    }

    setDays(num) {
        num !== 0 ? this.setState({daySelected: num}) : this.setState({daySelected : '0'});
    }

    setHours(num) {
        num !== 0 ? this.setState({hourSelected: num}) : this.setState({hourSelected : '0'});

    }

    setMinutes(num) {
        num !== 0 ? this.setState({minSelected: num}) : this.setState({minSelected : '0'});
    }


    //closer for modal
    close(){
        this.setState({ showModal: false });
    }

    //opener for modal
    open(course){
        //gets teacher recommendation and updates state
        let that = this;
        let coursesRef = ref.child("/courses/" + this.state.courseID + "/weekly");
        coursesRef.on("value", function(snapshot){
            that.setState({
                teacherRecommendation: getDaysHoursMins(snapshot.val()),
            });
        });
        this.setState({ 
            showModal: true,

        });
    }

    //helper returning h4 displaying current teacher recommendation
    displayTeacherRecommendation(){
        if(this.state.teacherRecommendation !== undefined){
            return (<h4>Current recommendation: {this.state.teacherRecommendation[0]} days, {this.state.teacherRecommendation[1]} hours, {this.state.teacherRecommendation[2]} minutes.</h4>)
        }else{
            return (<h4>No current recommendation set </h4>)
        }
    }

    render() {
        return (
            <div style={{marginBottom: '10px'}}>
                <Button block bsStyle="primary" onClick={this.open}><h4>Set recommended weekly time</h4></Button>

                <Modal className="modal" show={this.state.showModal} onHide={this.close} container={this}>
                    <Modal.Header closeButton>
                        <Form inline>
                            <FormGroup className="validation-form">
                                <Label className="bold">{this.state.courseID}</Label>
                                <Label className="plain-label">Enter recommended time spent per week:</Label>
                            </FormGroup>
                        </Form>
                    </Modal.Header>
                    <Modal.Body>
                        <div>
                            <p>Here you can set the recommended time use per week for students taking {this.state.courseID}.</p>
                            <p>Students will see your recommendation when setting their personal weekly goals.</p>
                            <p>One day is equal to 12 hours</p>
                            {this.getDropdowns(this.state.view)}
                            {this.displayTeacherRecommendation()}
                        </div>
                        <FeedbackMessage active={this.state.displayFeedbackMessage} message={this.state.feedbackMessage} bsStyle={this.state.bsStyle}/>
                    </Modal.Body>
                    <Modal.Footer>
                        <div>
                            <Button onClick={this.handleSaveBtn} bsStyle="success">Save</Button>
                            <Button onClick={this.close}>Close</Button>
                        </div>
                    </Modal.Footer>
                </Modal>
            
                    
            </div>
        );
    }
}