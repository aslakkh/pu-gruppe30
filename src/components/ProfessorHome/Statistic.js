/**
 * Created by jan on 21/03/2017.
 */
