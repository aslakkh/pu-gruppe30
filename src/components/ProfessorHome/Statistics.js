/**
 * Created by jan on 21/03/2017.
 */
import React, { Component } from 'react'
import firebase from 'firebase';
import {Button} from 'react-bootstrap'

Date.prototype.getWeek = function() {

    // Create a copy of this date object
    var target  = new Date(this.valueOf());

    // ISO week date weeks start on monday, so correct the day number
    var dayNr   = (this.getDay() + 6) % 7;

    // Set the target to the thursday of this week so the
    // target date is in the right year
    target.setDate(target.getDate() - dayNr + 3);

    // ISO 8601 states that week 1 is the week with january 4th in it
    var jan4    = new Date(target.getFullYear(), 0, 4);

    // Number of days between target date and january 4th
    var dayDiff = (target - jan4) / 86400000;

    if(new Date(target.getFullYear(), 0, 1).getDay() < 5) {
        // Calculate week number: Week 1 (january 4th) plus the
        // number of weeks between target date and january 4th
        return 1 + Math.ceil(dayDiff / 7);
    }
    else {  // jan 4th is on the next week (so next week is week 1)
        return Math.ceil(dayDiff / 7);
    }
};
const options = {
    // Elements options apply to all of the options unless overridden in a dataset
    // In this case, we are setting the border of each bar to be 2px wide and green
    elements: {
        rectangle: {
            borderWidth: 5,
            borderColor: 'rgb(0, 255, 0)',
            borderSkipped: 'bottom'
        }
    },
    responsive: true,
    legend: {
        position: 'top'
    },
    title: {
        display: true,
        text: 'Chart.js Bar Chart'
    }
};



let BarChart = require("react-chartjs").Bar;

const {Bar} = require("react-chartjs");
export default class Statistics extends Component {
    constructor(props) {
        super(props);
        this.list2=[];
        this.labels=[];
        this.show=false;
        this.state = {
            show: false,
            done:false,
            courseID: this.props.courseID,
            daily:0,
            weekly:0,
            monthly:0,
            data:{
                labels: [],
                datasets: [
                    {
                        label: 'Time',
                        backgroundColor: "rgba(0,220,0,0.5)",
                        data: []
                    }
                    ]
            }

        };
        this.sessions =[]
    }


    componentWillMount(){



        this.getDataFromFirebase()
    }
    componentDidMount(){


    }

    componentWillUnmount(){
        this.setState({
            show:false
        })
    }
    /*
    get sessions from database and selects only for the specific course
     */
    getDataFromFirebase(){
        console.log(this.state.courseID);
        let that = this;
        let kurs = this.state.courseID;
        let ting = [];
        let messageRef = firebase.database().ref();
        messageRef.child('users/').on('child_added', function(snapshot) {
            if(snapshot.val().courses[kurs] != undefined && snapshot.val().courses[kurs]['sessions']) {
                ting.push(snapshot.val().courses[kurs]['sessions']);
                ting.forEach(function (session) {
                    Object.keys(session).map((session2) => {that.sessions.push({'key': parseInt(session2), 'time': session[session2]['time']});


                })
            }
                )
        }});

    }
    /*
    Sorting function
     */
    sortNumber(a,b) {
        return a - b;
    }
/*
formats the sessions to usable data for chart.js
 */
    handleClick2(){
        if(this.sessions.length >0 &&!this.state.done){
            var list = this.sessions;
            Object.keys(list).map((key) => {
               let day = new Date(list[key]["key"]);
                if(isNaN(this.list2[day.getWeek()])){
                   this.list2[day.getWeek()]=0;
                }
                this.labels.indexOf(day.getWeek()) === -1 ? this.labels.push(day.getWeek()) : console.log("This item already exists");
                this.list2[day.getWeek()]=this.list2[day.getWeek()]+list[key]["time"]
            });
            this.list2 = this.list2.map((x) => x/3600);
            this.list2 = this.list2.filter(function(n){ return n != undefined });

            this.state.data.labels=this.labels.sort(this.sortNumber);
            this.state.data.datasets[0].data=this.list2;//.slice(this.list2.length-11,this.list2.length-1)
            //this.state.data.labels=this.state.data.labels;//.slice(this.state.data.labels.length-11,this.state.data.labels.length-1)
        this.setState({
            show:true,
            done:true
        })}
    }


    render() {
        if(!this.sessions.length > 0 || this.state.courseID === null || this.state.courseID === undefined){
            return (
                <h1>Loading2</h1>

            )}
            else{
            return(<div>
                    <Button bsStyle="primary" bsSize="large" onClick={this.handleClick2.bind(this)}>Get Statistics</Button>

                    {this.state.show ? <BarChart data={this.state.data}  options={options}/> : <h5>Loading statistics may take a while.</h5>}

                </div>
            )

        }
    }

}