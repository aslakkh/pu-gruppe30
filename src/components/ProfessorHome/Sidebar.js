import React, { Component } from 'react'
import {Tab,NavItem, Nav, Col,Row} from 'react-bootstrap'
import AddMessage from './AddMessage'
import MessageList from './MessageList'
import Statistics from './Statistics'
import SetTime from './SetTime'
import {
    BrowserRouter as Router,
    Route,
    Link,
    NavLink
} from 'react-router-dom'
import {LinkContainer} from 'react-router-bootstrap'

/*Maps the courses to make the sidebar and the site for the specific courses
 *
 *
 * Changes:
 * Change the way SpecificCourse is called, so you only load the course you are viewing.
 *
 * */

export default class Courses extends Component {
constructor(props) {
  super(props);
  this.state={
      courses: this.props.courses,
		}
  this.handleRoute = this.handleRoute.bind(this)
};

componentWillReceiveProps(nextProps){
  this.setState({
    courses: nextProps.courses
  });
}
    //matches URl (match.params.id) to course
    handleRoute({match}){
        return(
            <div>
                <AddMessage courseID={match.params.id}/>
                <MessageList courseID={match.params.id}/>
                <SetTime courseID={match.params.id} />
                <Statistics courseID={match.params.id}/>
            </div>
        )
    }
    
componentWillMount(){
}
	render (){

    if(this.state.courses === undefined || this.state.courses === null){ //checks if props are loaded
      return(
        <h4>No courses</h4>
      );
    }
    else{
      return (
          <Router>
              <div>
                  <Tab.Container id="left-tabs" >
                      <Row className="clearfix">
                          <Col sm={3}>
                              <Nav bsStyle="pills" className="navbar navbar-inverse" stacked >
                                  {Object.keys(this.state.courses).map((emne,i) =>


                                          <NavItem key={i} eventKey={i} className="navbar-link">
                                            <Link to={"/"+emne} key={i}>
                                                <div style={{marginTop: '-0.3em', fontSize:'1.5em',marginBottom: '-0.3em', color:'darkgrey'}}>
                                                {emne}
                                                </div>
                                            </Link>
                                    </NavItem>
                                  )}
                              </Nav>
                          </Col>
                          <Col sm={9}>
                              <Tab.Content animation>
                                  <Route path="/:id" render={this.handleRoute} />
                              </Tab.Content>
                          </Col>
                      </Row>
                  </Tab.Container>
              </div>
          </Router>
      )
    }
		
	}




}