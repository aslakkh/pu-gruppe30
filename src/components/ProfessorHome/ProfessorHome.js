/**
 * Created by jan on 18/03/2017.
 */
import React, { Component } from 'react'
import Sidebar from './Sidebar'

export default class ProfessorHome extends Component {
    constructor(props) {
        super(props);
        this.state = {
            courses: this.props.courses

        };
    }

    getData(){

    }

    castCommentToFirebase(){

    }

    render() {
        return (
            <Sidebar courses={this.state.courses}/>

        )
    }

}