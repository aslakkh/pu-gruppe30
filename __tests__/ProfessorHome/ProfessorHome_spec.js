/**
 * Created by jan on 19/04/2017.
 */
import React from 'react';
import renderer from 'react-test-renderer';
import ProfessorHome from '../../src/components/ProfessorHome/ProfessorHome.js';
import TestUtils from "react-addons-test-utils"

const courses={
    TES1000:{active: true,
        goals: {dailyGoal:{active: false,
            timeSet: 0,
            timeSpent: 0,
            value: 0},
            monthlyGoal:
                {active:false,
                    timeSet:0,
                    timeSpent:0,
                    value:0},
            weeklyGoal:
                {active:false,
                    timeSet:0,
                    timeSpent:0,
                    value:0}

        }
    }};
const coursesNull=null;

describe("Home", function () {
    let renderer2;
    let renderer3;
    let renderer4;
    let component2;
    let component3;
    let component4;
    renderer2 = TestUtils.createRenderer();
    renderer2.render(
        <ProfessorHome courses={courses}/>
    );
    component2 = renderer2.getRenderOutput();
    it("Not logged in", () => {
        expect(component2).toMatchSnapshot();
    });

});