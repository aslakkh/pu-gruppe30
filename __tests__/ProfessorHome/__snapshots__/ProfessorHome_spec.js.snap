// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Not logged in 1`] = `
<Courses
  courses={undefined}
/>
`;
